File: scripts/ops/release-check-version/create_components_json.py
import os

import requests
import sys
import json


token = os.environ.get("GITHUB_TOKEN")
if not token:
    print("GITHUB_TOKEN is not set.")
    sys.exit(1)

headers = {"Authorization": f"token {token}"} if token else {}


def parse_release_version(branch):
    prefix = "release-"
    if not branch.startswith(prefix):
        return None
    return branch[len(prefix):].split("-", 1)[0]


def is_branch_before_release(branch, target_version):
    version_str = parse_release_version(branch)
    if version_str is None:
        return False

    target_version_str = ".".join(map(str, target_version))
    try:
        from packaging.version import parse, InvalidVersion
        return parse(version_str) < parse(target_version_str)
    except (ImportError, InvalidVersion):
        return False


def get_latest_commit_hash(repo, branch):
    url = f"https://api.github.com/repos/{repo}/commits/{branch}"
    response = requests.get(url, headers=headers)
    if response.status_code == 200:
        return response.json()['sha']
    else:
        print(f"Error fetching {repo} branch {branch}: {response.status_code}")
        return None


def main(branch, version):
    TICDC_REPO_CHANGE_VERSION = (8, 5)
    ticdc_repo = "pingcap/tiflow" if is_branch_before_release(branch, TICDC_REPO_CHANGE_VERSION) else "pingcap/ticdc"

    repos = {
        "binlog": "pingcap/tidb-binlog",
        "br": "pingcap/tidb",
        "tidb": "pingcap/tidb",
        "tikv": "tikv/tikv",
        "pd": "tikv/pd",
        "tiflash": "pingcap/tiflash",
        "ticdc": ticdc_repo,
        "dm": "pingcap/tiflow",
        "dumpling": "pingcap/tidb",
        "tidb-dashboard": "pingcap/tidb-dashboard",
        "lightning": "pingcap/tidb",
        "ng-monitoring": "pingcap/ng-monitoring",
        "enterprise-plugin": "pingcap-inc/enterprise-plugin"
    }

    output = {"docker_images": [], "tiup_packages": []}

    for name, repo in repos.items():
        commit_hash = get_latest_commit_hash(repo, branch)
        if commit_hash:
            entry = {"name": name, "version": version, "commit_hash": commit_hash}
            output["docker_images"].append(entry)
            output["tiup_packages"].append(entry)  # Assuming you want the same for tiup_packages

    with open(f"components-{version}.json", "w") as f:
        json.dump(output, f, indent=2)


if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Usage: create_components_json.py <branch> <version>")
        sys.exit(1)
    branch, version = sys.argv[1], sys.argv[2]
    main(branch, version)

File: tools/deprecated/release-check-version/create_components_json.ts
#!/usr/bin/env deno run --allow-net --allow-env --allow-write

const token = Deno.env.get("GITHUB_TOKEN");
if (!token) {
  console.error("GITHUB_TOKEN is not set.");
  Deno.exit(1);
}

const headers = { "Authorization": `token ${token}` };

function parseReleaseVersion(branch: string): number[] | null {
  const prefix = "release-";
  if (!branch.startsWith(prefix)) return null;
  const versionPart = branch.substring(prefix.length).split("-", 1)[0];
  if (!versionPart) return null;
  const parts = versionPart.split(".").map((p) => {
    const n = parseInt(p, 10);
    return isNaN(n) ? NaN : n;
  });
  if (parts.length === 0 || parts.includes(NaN)) return null;
  return parts;
}

function isBranchBeforeRelease(
  branch: string,
  targetVersion: number[],
): boolean {
  const version = parseReleaseVersion(branch);
  if (!version) return false;
  const maxLen = Math.max(version.length, targetVersion.length);
  const paddedVersion = version.slice();
  while (paddedVersion.length < maxLen) paddedVersion.push(0);
  const paddedTarget = targetVersion.slice();
  while (paddedTarget.length < maxLen) paddedTarget.push(0);
  for (let i = 0; i < maxLen; i++) {
    if (paddedVersion[i] < paddedTarget[i]) return true;
    if (paddedVersion[i] > paddedTarget[i]) return false;
  }
  return false;
}

async function getLatestCommitHash(
  repo: string,
  branch: string,
): Promise<string | null> {
  const url = `https://api.github.com/repos/${repo}/commits/${branch}`;
  const resp = await fetch(url, { headers });
  if (resp.ok) {
    const data = await resp.json();
    return data.sha as string;
  } else {
    console.error(`Error fetching ${repo} branch ${branch}: ${resp.status}`);
    return null;
  }
}

async function main(branch: string, version: string): Promise<void> {
  const TICDC_REPO_CHANGE_VERSION = [8, 5];
  const ticdcRepo = isBranchBeforeRelease(branch, TICDC_REPO_CHANGE_VERSION)
    ? "pingcap/tiflow"
    : "pingcap/ticdc";

  const repos: Record<string, string> = {
    // "binlog": "pingcap/tidb-binlog",
    "br": "pingcap/tidb",
    "tidb": "pingcap/tidb",
    "tikv": "tikv/tikv",
    "pd": "tikv/pd",
    "tiflash": "pingcap/tiflash",
    "ticdc": ticdcRepo,
    "dm": "pingcap/tiflow",
    "dumpling": "pingcap/tidb",
    "tidb-dashboard": "pingcap/tidb-dashboard",
    "lightning": "pingcap/tidb",
    "ng-monitoring": "pingcap/ng-monitoring",
    "enterprise-plugin": "pingcap-inc/enterprise-plugin",
  };

  const output: { docker_images: any[]; tiup_packages: any[] } = {
    docker_images: [],
    tiup_packages: [],
  };

  for (const [name, repo] of Object.entries(repos)) {
    const commitHash = await getLatestCommitHash(repo, branch);
    if (commitHash) {
      const entry = { name, version, commit_hash: commitHash };
      output.docker_images.push(entry);
      output.tiup_packages.push(entry); // same entry for tiup_packages
    }
  }

  const filename = `components-${version}.json`;
  await Deno.writeTextFile(filename, JSON.stringify(output, null, 2));
}

if (import.meta.main) {
  const args = Deno.args;
  if (args.length !== 2) {
    console.error("Usage: create_components_json.ts <branch> <version>");
    Deno.exit(1);
  }
  const [branch, version] = args;
  await main(branch, version);
}
